File: challenges/project-euler/maximum_path_sum_i.js
var fs = require('fs');
var triangleArray = []

function readLines(input, func) {
  var remaining = '';
  var lineIndex = 0;
  input.on('data', function(data) {
    remaining += data;
    var index = remaining.indexOf('\n');
    while (index > -1) {
      var line = remaining.substring(0, index);
      triangleArray.push(line.split(" ").map(Number))
      remaining = remaining.substring(index + 1);
      // func(line, lineIndex);
      index = remaining.indexOf('\n');
      lineIndex++;
    }
  });

  input.on('end', function() {
    console.log(triangleArray)
    console.log('end')
    if (remaining.length > 0) {
      // func(remaining);
    }
  });
}

function func(data, index) {
  console.log('Line ' + index + ":" + data);
  triangleArray.push(data)
}

var input = fs.createReadStream('data/maximum_path_sum_i_data.txt');
readLines(input, func);


var node = function(value){
  this.value = value;
  this.rightChild = null;
  this.leftChild = null;
}

var nodeTree = function(){
  this.root = null;
  this.size = null;
}

nodeTree.prototype.insertValueOnRow = function(val, row) {

}

File: challenges/project-euler/maximum_path_sum_i.rb
triangle_array = []

File.open('data/p067_triangle.txt', "r") do |f|
  f.each_line do |line|
    triangle_array.push(line.split(" ").map {|num| {num: num.to_i } })
  end
end

triangle_array.each_with_index do |row, row_index|
  row.each_with_index do |num, col_index|
    # add
    next_row = triangle_array[row_index + 1]
    if next_row
      num[:left] = next_row[col_index]
      num[:right] = next_row[col_index + 1]
    end
  end
end

root = triangle_array[0][0]

def walk_tree(node)
  # two base cases
  if node.nil?
    return 0
  end
  if node.key?('memo_sum')
    return node[:memo_sum]
  end
  # calculate sum
  node[:memo_sum] = node[:num] + [walk_tree(node[:left]),walk_tree(node[:right])].max
  return node[:memo_sum]
end

p walk_tree(root)
